service/src/main/java/edu/tamu/catalog/controller/PatronController.java

```diff
@@ -0,0 +1,54 @@
+package edu.tamu.catalog.controller;
+
+import java.text.ParseException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import edu.tamu.catalog.domain.model.FeesFines;
+import edu.tamu.catalog.exception.NotFoundException;
+import edu.tamu.catalog.service.CatalogService;
+import edu.tamu.catalog.service.CatalogServiceFactory;
+
+@RestController
+@RequestMapping("/patron")
+public class PatronController {
+
+    @Autowired
+    private CatalogServiceFactory catalogServiceFactory;
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    /**
+     * Provides data for all fees and fines associated with a patron.
+     *
+     * @param String catalogName (optional)
+     * @param String user
+     * @return
+     * @throws ParseException
+     */
+    @GetMapping("/{uin}/fines")
+    public @ResponseBody ResponseEntity<FeesFines> fines(@RequestParam(value="catalogName", defaultValue="folio") String catalogName, @PathVariable String uin) throws ParseException {
+        FeesFines feesFines = getCatalogServiceByName(catalogName).getFeesFines(uin);
+
+        if (feesFines == null) {
```

Comment:
This never occurs.

Reply:
This occurs whenever the query parameter `catalogName` is set to something like `evans`.
That catalog calls the VoyagerCatalogService, [which only returns null for getFees()](https://github.com/TAMULib/CatalogService/blob/572aa52cac87801bf49b6c6dedebf285d82c2b7e/service/src/main/java/edu/tamu/catalog/service/VoyagerCatalogService.java#L287).
If the Catalog doesn't support the action, it returns NULL and is caught by this condition.

Reply:
I see. That is why we should throw [NotImplementedException](https://commons.apache.org/proper/commons-lang/javadocs/api-release/org/apache/commons/lang3/NotImplementedException.html). It should have an exception handler and return the appropriate status code. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501

Reply:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.2

Reply:
402 responses in these cases are helpful to determine the problem with the request.

Reply:
The problem with 501 is its use of the word **Method** as in [Request Method](https://tools.ietf.org/html/rfc2616#section-5.1.1).

As for your use here, I think a **405 Method Not Allowed** is more appropriate because: "_An origin server SHOULD return the status code 405 (Method Not Allowed) if the method is known by the origin server but not allowed for the  requested resource..._"

The problem is, I have seen browsers interpret **Method** here as purely with **GET**, **POST**, etc..
The browser then make assumptions and weird problems may happen.

Officially the standard mentions the following under https://tools.ietf.org/html/rfc2616#section-9 :
```
   In particular, the convention has been established that the GET and
   HEAD methods SHOULD NOT have the significance of taking an action
   other than retrieval. These methods ought to be considered "safe".
   This allows user agents to represent other methods, such as POST, PUT
   and DELETE, in a special way, so that the user is made aware of the
   fact that a possibly unsafe action is being requested.
```

This was a few years ago, so maybe browsers have changed how they interpret the use of the word **Method**.
The standard could improve their wording because as it stands.

We should be "safe", but outside of HEAD requests, browsers have (historically) badly handled 405/501 outside of HEAD request responses.

All that said, I am in favor or 405 (over 501) due to its semantic correctness as Voyager should never support this specific endpoint.
Unimplemented suggests that it will be implemented in the future.

Reply:
405 Method Not Allowed is for REST Method Not Allowed. https://tools.ietf.org/html/rfc7231#section-4

Reply:
405 response would include a list of accepted REST methods for the given endpoint. This is clearly not what we have. We have an interface that has an unsupported feature for a given query parameter.

Reply:
> When a request method is received
   that is unrecognized or not implemented by an origin server, the
   origin server SHOULD respond with the 501 (Not Implemented) status
   code.  When a request method is received that is known by an origin
   server but not allowed for the target resource, the origin server
   SHOULD respond with the 405 (Method Not Allowed) status code.

Reply:
We do recognize the request but it is not implemented. `not implemented by an origin server`

Reply:
If we are concerned with the semantic description of 501, I would just go with 500.

Reply:
What I was getting at is that [both the 501 and the 405 are specifically for Request Method processing](https://tools.ietf.org/html/rfc7231#section-4).
```   The set of methods allowed by a target resource can be listed in an
   Allow header field (Section 7.4.1).  However, the set of allowed
   methods can change dynamically.  When a request method is received
   that is unrecognized or not implemented by an origin server, the
   origin server SHOULD respond with the 501 (Not Implemented) status
   code.  When a request method is received that is known by an origin
   server but not allowed for the target resource, the origin server
   SHOULD respond with the 405 (Method Not Allowed) status code.
```
This block is about responses to the HTTP HEAD method.
So if a 501 is to be used then a 405 is also just as valid (or invalid).

This part is of importance in the quote:
```
..When a request method is received that is known by an origin
server but not allowed for the target resource..
```

Our endpoint is known, but is not allowed for the Voyager catalog.
```
..is known by an origin server but not allowed..
```

Further, https://tools.ietf.org/html/rfc7231#section-6.6.2 :
```
   The 501 (Not Implemented) status code indicates that the server does
   not support the functionality required to fulfill the request.  This
   is the appropriate response when the server does not recognize the
   request method and is not capable of supporting it for any resource.

   A 501 response is cacheable by default; i.e., unless otherwise
   indicated by the method definition or explicit cache controls (see
   Section 4.2.2 of [RFC7234]).
```
   I cannot emphasize enough that the 501, like 405, are specifically about the **Request Method** part and not about the **Response** as a whole.
   Essentially, returning a 501 if **HTTP DELETE** is not implemented or not recognized; however, if it is not allowed but is known, a 405 would be returned for the **HTTP DELETE**. The difference between 405 and 501 is primarily context. (in a similar way to how temporary redirects and permanent redirects differ.)

Reply:
That all makes sense except for the semantic of allowed. Allowing something makes me think of that as authorization. 

I agree, neither fit it perfectly. Maybe we should just use 500.

Reply:
Either 500 or 400. 402 Not Found least accurate for these endpoints.

Reply:
We could possibly return a `204 No Content`. There are again, problems with the vaguery of their definitions.
But I prefer an error in this case (and I worry that browsers also handle 204 only in certain conditions).

I think I can settle with **400** because **500** is just wrong here (and I would say more wrong than a **404**).
```
   ...status code indicates that the server cannot or
   will not process the request due to something that is perceived to be
   a client error...
```

"_Will not process the request due to something that is perceived to be a client error._"
I think that those words satisfy our use case.
"Its the clients fault for trying to get a FOLIO specific thing from a Voyager specific endpoint".
ergo, a client error.

Reply:
I would not want a 200 code for this. If for some reason we are making requests using Voyager catalog, we need a response that indicates it needs to change. I can see 200 No Content being confused here.

We are likely never going to implement the Voyager implementation for these and want the response to indicate **do not use**. 400 Bad Request should suffice. I was thinking a 500 is a bit more severe but doesn't necessarily inform the client to change the request.